vm/src/value.rs

```diff
@@ -84,24 +84,238 @@ impl<'vm, ENV> Iterator for PairIter<'vm, ENV> {
     }
 }
 
-// Do this wrap nonsense so that Value is hashable...
-#[derive(Copy, Clone, Debug)]
-pub struct F32Wrap(pub f32);
-
-impl PartialEq for F32Wrap {
+/// This struct uses 7 bytes to represent a floating point number.
+/// Most operations are done by converting to f64, performing the operation, and then converting back to F56
+///
+/// F56 uses 1 bit for the sign, 10 bits for the exponent, and 45 bits for the mantissa.
+/// Compared to f32, it has +2 exponent bits and +22 mantissa bits.
+/// Compared to f64, it has -1 exponent bit and -7 mantissa bits.
+///   Byte 0    Byte 1    Byte 2    Byte 3    Byte 4    Byte 5    Byte 6
+/// [sEEEEEEE][EEEmmmmm][mmmmmmmm][mmmmmmmm][mmmmmmmm][mmmmmmmm][mmmmmmmm]
+///
+/// Exponent bits range from 0 to 1023
+/// they represent -511 to +512 but are stored biased by +511
+/// the exponent of -511 is reserved for the number 0 and subnormal numbers
+/// the exponent of +512 is reserved for infinity and NaN
+/// so normal exponents range from -510 to +511
+///
+/// smallest positive subnormal value is 8.48e-168 (2^-555)
+/// smallest positive normal value is 2.98e-154 (2^-510)
+/// maximum finite value is 1.34e154
+#[derive(Copy, Clone)]
+pub struct F56(pub [u8; 7]);
+impl Eq for F56 {}
+impl PartialEq for F56 {
     fn eq(&self, other: &Self) -> bool {
-        (self.0 - other.0).abs() < f32::EPSILON
-        //self.0.to_bits() == other.0.to_bits()
+        // Allow NaN == NaN so equality is reflexive and we can impl Eq to use F56 as a hash key
+        if f64::from(*self).is_nan() && f64::from(*other).is_nan() {
```

Comment:
You are calling f64::from() twice, once here and again below.  Recommend you move the let value and let other_value statements up and only call the from once per value vs twice.  The from is not super costly but it is not free.